### PR TITLE
Finalize CI release API integration

### PR DESCRIPTION
**Pull request by justinas**
Contributes to: https://github.com/gravitational/cloud/issues/1553

* Switch to the production instance of the release server
* Stop ignoring failures in the "register artifacts" / "publish release" steps

### CODE REVIEW

**Review by codingllama — APPROVED** [review]

File: .drone.yml [hunk 1]
```diff
@@ -949,11 +949,11 @@ steps:
   - Get-Relcli -Url $relcliUrl -Sha256 $relcliSha256 -Workspace $Workspace
   - Register-Artifacts -Workspace $Workspace -Outputs $OutputsDir
   environment:
-    RELCLI_BASE_URL: https://releases-staging.platform.teleport.sh
```
[the remark this answers is not in the document]
Comment (justinas) [reply .drone.yml:952]: The `publish-rlz` pipeline is not dronegen-d, so its changes are manual, others are all generated.

**Review by avatus — APPROVED** [review]

File: .drone.yml [hunk 2]
```diff
@@ -1846,10 +1846,9 @@ steps:
     done
   environment:
     RELEASES_CERT:
-      from_secret: RELEASES_CERT_STAGING
+      from_secret: RELEASES_CERT
     RELEASES_KEY:
-      from_secret: RELEASES_KEY_STAGING
-  failure: ignore
```
Comment (r0mant) [inline .drone.yml:1852]: If we stop ignoring failures, doesn't this mean that every subsequent rerun of the tag build will fail? Right now "publish to release service" step fails every time we re-run tag build (e.g. due to a failure).

**Review by r0mant — CHANGES_REQUESTED** [review]

File: .drone.yml (hunk 2 above)
Comment (justinas) [reply .drone.yml:1852]: Yes. The solution for now is to manually destroy the partially-created release, and then retry the build. We have [created a runbook for this](https://github.com/gravitational/cloud/blob/master/docs/runbooks/releases.md#destroy-a-release) and noted this in the [release checklist](https://gravitational.slab.com/posts/teleport-release-checklist-cf0s6vmt#h9b88-release-process) (Ctrl-F "If the tag build failed").

Comment (r0mant) [reply .drone.yml:1852]: What is the plan to address this limitation and what is the ETA?

I publish 99% of Teleport releases and it's already a pretty involved process so I would really love if there was no yet another thing that I have to do manually.

Comment (justinas) [reply .drone.yml:1852]: One straightforward solution comes to mind - perhaps Drone could execute this "destroy" workflow before reretrying the tag build. I'll try to prototype this ASAP, and if it works, perhaps we can slip in that change before this PR.

Comment (justinas) [reply .drone.yml:1852]: @r0mant Please see https://github.com/gravitational/teleport/pull/16669 . If the solution is sufficient, I will try to rebase / remake this PR on top of it.

**Comment by r0mant** [comment]
@justinas I just ran into another tricky situation trying to publish 10.2.3 which demonstrates the deficiencies in the current implementation:

- Tag build succeeded: https://drone.platform.teleport.sh/gravitational/teleport/15816.
- Promotion pipeline failed due to an issue with our AWS account (IT/sec team enabled encrypted AMIs so it failed trying to make them public): https://drone.platform.teleport.sh/gravitational/teleport/15818.
- That failure required me to re-run the tag build again to rebuild AMIs, which failed trying to register artifacts for one of the assets so it didn't reach the AMI step: https://drone.platform.teleport.sh/gravitational/teleport/15821/1/8. Seems we don't ignore the release service errors there like we do in other pipelines.

So now we ended up with a partially published release which I'm not sure how to complete. I looked at the `relcli` instructions but it only appears to provide a way to destroy the entire release - I'm not sure what this means, all artifacts? Is there a way to unpublish a single artifact? I don't want to delete all artifacts because we already shared them with a customer who needs a 10.2.3 patch.

Our release API should be able to handle this scenario without manual intervention so we can re-run build/publish jobs in situations like this. I think I suggested this before - can we handle "already exists" error in the "register artifacts" steps gracefully and not fail the entire pipeline? Then we won't need to destroy anything prior to the pipeline run.

**Comment by justinas** [comment]
@r0mant Hm, that is definitely tricky. I can't say I imagined this workflow existing at all (build; publish (failed); build; publish). This is also why https://github.com/gravitational/teleport/pull/16669 is the way it is - I imagined we would only rebuild when it is still *possible* to destroy the artifacts, i.e. in the `draft` the stage of the release.

> which failed trying to register artifacts for one of the assets so it didn't reach the AMI step

This is very strange to me, I was under the impression that `failure: ignore` would mean that dependent steps would continue running. It seems not. :confused: 

> Is there a way to unpublish a single artifact?

For published releases, no. This was one of the main design principles:

> It will not be possible to re-attach/change the asset of the `published` releases.
> https://github.com/gravitational/cloud/blob/master/rfd/0004-Release-Asset-Management.md#to-attach-an-asset-to-the-release

I would argue this is somewhat similar to how we do not re-tag in Git once a vX.Y.Z has been tagged, even if it fails (we usually go to vX.Y.Z+1 if we can't fix the given release in my understanding). Once a build has been exposed to customers, the release API does not allow to pull the rug from underneath and replace the artifacts with something else.

---

> I think I suggested this before - can we handle "already exists" error in the "register artifacts" steps gracefully and not fail the entire pipeline?

I think implementing this check and skipping the re-upload of artifacts would be fine, if the release has already been published (although I would really like to push for the [full move to relcli first](https://github.com/gravitational/cloud/issues/2250), as implementing complicated logic in shell scripts is tricky).

A blanket "ignore (skip re-upload) if exists" is IMO not an ideal solution:
* It is somewhat unintuitive to retry the build pipeline, have it rebuild the assets, and then essentially do nothing with them (silently). This will be highly confusing in another "CI is technically green, but I still want to re-make artifacts" situation. 
* It is an approach inconsistent with artifacts outside of the CDN:
  * AFAIK deb/rpm repos are built from `clientbuilds` still, which is implicitly overwritten after a rebuild.
    * This would result in a different `teleport-vX.Y.Z.deb` in CDN vs DEB repo, etc. 
  * From your description, AMIs seem to behave the same way (silently overwrite at any time).

Perhaps a better way to handle this would be to avoid *rebuilding* (not just re-uploading) at all in the individual pipelines, if that particular pipeline was successful (even if others in the entire run was not). I am not aware of such functionality to "partially re-run" in Drone, so this probably means that the build process itself would need to be made smarter. That is, the early steps of the pipeline would need to be aware of what will be produced at the end, and whether these artifacts already exist (e.g. using the release API as the source of truth), and thus whether the build can be skipped.

---

Looking at this in a very pragmatic way for a second, it would be interesting to guesstimate the odds of the next failure being:
* An intermittent error in the "build" phase (intermittent network error or similar), which would be handled by the proposed implementation
vs.
* An error like the latest one, caused by background changes to the infra, *and* triggered by something in the "build" phase, *and* only materializing in the "promote" phase, *and* only fixable by going back and re-trying the "build" phase with no code changes

**Comment by justinas** [comment]
Now that https://github.com/gravitational/teleport/pull/16669 has been merged, I will try to rebase / remake this PR entirely on top of that.

**Comment by justinas** [comment]
Closed in favor of #16834

**Pull request closed without merging**